meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java

```diff
@@ -0,0 +1,136 @@
+package org.apache.helix.metaclient.recipes.lock;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.helix.metaclient.api.MetaClientInterface;
+import org.apache.helix.metaclient.datamodel.DataRecord;
+import org.apache.helix.metaclient.factories.MetaClientConfig;
+import java.util.Collection;
+import java.util.concurrent.TimeUnit;
+
+public class DistributedSemaphore {
+
+  /**
+   * Create a distributed semaphore client and creates a semaphore at the indicated path with the given capacity.
+   */
+  public DistributedSemaphore(MetaClientConfig config, String path, int capacity) {
+    throw new NotImplementedException("Not implemented yet.");
+  }
+
+  /**
+   * Create a distributed semaphore client. Connects to an existing semaphore at given path.
+   */
+  public DistributedSemaphore(MetaClientConfig config, String path) {
+    throw new NotImplementedException("Not implemented yet.");
+  }
+
+  /**
+   * Connect to an existing distributed semaphore client and semaphore at the indicated path.
+   */
+  public DistributedSemaphore(MetaClientInterface<DataRecord> client, String path) {
```

Comment:
We should have a constructor that also takes an existing client for producer as well?

Reply:
I see...in this case it would make sense to have a create and retrieve semaphore methods separate to that of the constructor.